[EZP-25721] Warn devs upfront of ConfigResolver initialization order issues (#634)
ezsystems/developer-documentation: docs/guide/config_dynamic.md

// $myParamSetting's value will be 'another value'
```

In debug mode, ConfigResolver detects if parameters were loaded prior to initialization of SiteAccess to warn about issues that are hard to resolve.
It will log all instances of `getParameter()` that may be used unsafely.

> Additionally, for class-based services, the warning appears even if a service is recognized by `DynamicSettingsListener` as `$updatableServices` (which is safe).

Reviewer: I think we can skip this, it’s implied in warning on usages deemed unsafe

- Avoid eager usage of config resolver (e.g. in service factories).

> - Instead of using `ctor('$dynamic_param$')`, use `(setter('$dynamic_param$'))` as it allows the system to update the scope changes.

Reviewer: ```suggestion
- Instead of using `ctor('$dynamic_param$')`, use `(setter('$dynamic_param$'))` as it allows the system to update your service with changes on scope changes.
```

> - Do not use ConfigResolver in service factory.

Reviewer: Hmm, this can be misleading.
Where did you take this from?

Author: I picked it from [this comment](https://github.com/ezsystems/ezpublish-kernel/pull/2636/files#diff-26112d3c76ce033446c6bf175dbdae01R323), but I may have misinterpreted something. Your suggestion would be to delete this line?

Reviewer: So maybe then `Try to avoid eager usage of config resolver (e.g. in service factories)`

> - Load the parameter lazily by injecting ConfigResolver with the required value only when necessary.

Reviewer: ```suggestion
- Load the parameter lazily by injecting ConfigResolver, and get the parameter from it _when_ you need to instead of during construction.
```

> - Instead of using `ctor('$dynamic_param$')`, use `(setter('$dynamic_param$'))` as it allows the system to update the scope changes or use ConfigResolver.

Reviewer: “Or use config resolver”
This can be own point about rather injecting config resolver and only using the value needed when necessary, aka load the parameter lazily

> - Try using [lazy commands](https://symfony.com/doc/current/console/lazy_commands.html) (applicable only to eZ Platform v2.5 or higher).

Reviewer: ```suggestion
- Try using [lazy commands](https://symfony.com/doc/current/console/lazy_commands.html) (applicable only to Symfony v3.4 or higher).
```

Unsure about this one, but more precise at least. Up to you.

> - You can also use ConfigResolver lazily by [making commands lazy](https://symfony.com/doc/current/console/lazy_commands.html) or configure [lazy services](https://symfony.com/doc/current/service_container/lazy_services.html).

Reviewer: You can skip configresolver here, it’s not that relevant as it might be indirect usage (you might depend on a regular service, which in turn depends on configresolver)

So somehow state you can avoid the issue by either making command lazy (2.5+ only) or service lazy (does not apply to commands)

2. `$namespace` - your application namespace, `myapp` in the previous example. If null, the default namespace will be used, which is `ezsettings` by default.
3. `$scope` - a SiteAccess name. If null, the current SiteAccess will be used.

!!! note

    In debug mode, ConfigResolver detects if parameters were loaded prior to initialization of SiteAccess to warn about issues.
    It will log all instances of `getParameter()` that may be used unsafely.
    

>     If a problem occurs, you will reveive a warning that ConfigResolver had been used to load parameter `languages`

Reviewer 2: ```suggestion
    If a problem occurs, you will receive a warning that ConfigResolver had been used to load parameter `languages`
```

>     before SiteAccess was loaded by services: `my.own.service` and `ez.service.used.to.early`.

Reviewer 2: ```suggestion
    before SiteAccess was loaded by services: `my.own.service` and `ez.service.used.too.early`.
```

2. `$namespace` - your application namespace, `myapp` in the previous example. If null, the default namespace will be used, which is `ezsettings` by default.
3. `$scope` - a SiteAccess name. If null, the current SiteAccess will be used.

> ### ConfigResolver unsafe parameters  

Author: That's a great idea!
